---
title: "Multiplexing Software Benchmark"
output:
  html_document:
    df_print: paged
    toc: true
    theme: 'yeti'
    highlight: 'tango'
    code_folding: hide
params:
    rmd: NULL
    dataset: NULL
    Demuxlet_V1: NULL
    Demuxlet_V2: NULL
    SoC: NULL
    Vireo: NULL
    SCanSNP: NULL
    dropkick: NULL
    scDblFinder: NULL
    BarcodeMap: NULL
    FilteredFeaturesPath: NULL
    metricsFile: NULL
    aggregatedCall: NULL
---

```{r, collapse=TRUE}





Dataset <- snakemake@params[["dataset"]]
Demuxlet_V1_path <- snakemake@input[["demuxlet_v1"]]
Demuxlet_path <- snakemake@input[["demuxlet_v2"]]
SoC_path <- snakemake@input[["soc"]]
Vireo_path <- snakemake@input[["vireo"]]
SCanSNP_path <- snakemake@input[["scansnp"]]
dropkick_path <- snakemake@input[["dropkick"]]
scDblFinder_path <- snakemake@input[["scDblFinder"]]
BarcodeMap <- snakemake@params[["barcode_map"]]
FilteredFeaturesPath <- snakemake@params[["filtered_matrix"]]
metricsFile <- snakemake@params[["demult_metrics"]]
aggregatedCall <- snakemake@params[["aggregatedCall"]]
aggregatedCalldir <- snakemake@params[["aggregatedCalldir"]]

```
####  Analysis is performed on `r Dataset`.

## __1. SetUp__


```{r, message = FALSE, collapse=TRUE}
library(tidyr)
library(combinat)
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(grid)
library(Seurat)
library(VennDiagram)


options(stringsAsFactors = FALSE)

dir.create(file.path(aggregatedCalldir), showWarnings = FALSE)

```




****

## __2. Read data__

 * Soup or Cell
 * Vireo
 * Demuxlet V1
 * Demuxlet V2
 * SCanSnp
 * scDblFinder



```{r ReadData,, collapse=TRUE}
# SoC
SoC <- read.csv(SoC_path, header = T, sep = "\t",stringsAsFactors=F)
SoC$SoCID <- ifelse(SoC$status == "singlet", SoC$assignment, SoC$status)

# Vireo
Vireo <- read.csv(Vireo_path, header = T, sep = "\t",stringsAsFactors=F) %>% dplyr::rename(barcode=cell, VireoID=donor_id)
Vireo <-  Vireo[,c("barcode","VireoID")]

# Demuxlet V1
Demuxlet_V1 <- read.csv(Demuxlet_V1_path, header = T, sep = "\t",stringsAsFactors=F) %>% dplyr::rename(barcode=BARCODE) %>%
  tidyr::separate(BEST, c('dbltype', 'first','second'), sep="-")
Demuxlet_V1$Demuxlet_V1_BEST <- ifelse(Demuxlet_V1$dbltype != "SNG", Demuxlet_V1$dbltype, Demuxlet_V1$first)
Demuxlet_V1[Demuxlet_V1=="DBL"]<-"doublet"
Demuxlet_V1[Demuxlet_V1=="AMB"]<-"unassigned"


# Demuxlet V2 (henceforth Demuxlet.)
Demuxlet <- read.csv(Demuxlet_path, header = T, sep = "\t",stringsAsFactors=F)  %>% dplyr::rename(barcode=BARCODE) %>%
  tidyr::separate(BEST.GUESS, c('BEST.GUESS_1', 'BEST.GUESS_2','Prop'), sep=",")
Demuxlet$Demuxlet_BEST <- ifelse(Demuxlet$DROPLET.TYPE != "SNG", Demuxlet$DROPLET.TYPE, Demuxlet$BEST.GUESS_1)
Demuxlet[Demuxlet=="DBL"]<-"doublet"
Demuxlet[Demuxlet=="AMB"]<-"unassigned"

# ScanSnp
Scansnp <- read.csv(SCanSNP_path, sep = "\t", header = T,stringsAsFactors=F) %>% dplyr::rename(SCanSNPID=ID)
Scansnp$SCanSNPID[Scansnp$SCanSNPID == "notclassified"] <- "unassigned"
Scansnp[Scansnp=="Doublet"]<-"doublet"

# scdblfinder
if (!is.null(FilteredFeaturesPath)){
  scDblFinder <- read.csv(scDblFinder_path, sep = "\t", header = F,stringsAsFactors=F,
  col.names=c("barcode","scDblFinder_type"))
}


#dropkick
if (!is.null(FilteredFeaturesPath)){
  dropkick <- read.csv(dropkick_path, sep = "\t", header = T,stringsAsFactors=F)
}


```



```{r MergeData, collapse=TRUE}

if (!is.null(FilteredFeaturesPath)){
mergedIDs <- dplyr::select(SoC, c("barcode","SoCID"))  %>%
  dplyr::inner_join(dplyr::select(Demuxlet_V1, c("barcode","Demuxlet_V1_BEST")),  by = "barcode") %>%
  dplyr::inner_join(dplyr::select(Vireo, c("barcode","VireoID")),  by = "barcode") %>%
  dplyr::inner_join(dplyr::select(Scansnp, c("barcode","SCanSNPID")),  by = "barcode") %>%
    dplyr::inner_join(dplyr::select(Demuxlet, c("barcode","Demuxlet_BEST")), by = "barcode") %>%
    dplyr::inner_join(dplyr::select(scDblFinder, c("barcode","scDblFinder_type")), by = "barcode") %>%
      dplyr::inner_join(dplyr::select(dropkick, c("barcode","dropkick_label_03")), by = "barcode")
  } else {
  mergedIDs <- dplyr::select(SoC, c("barcode","SoCID"))  %>%
    dplyr::inner_join(dplyr::select(Demuxlet_V1, c("barcode","Demuxlet_V1_BEST")),  by = "barcode") %>%
    dplyr::inner_join(dplyr::select(Vireo, c("barcode","VireoID")),  by = "barcode") %>%
    dplyr::inner_join(dplyr::select(Scansnp, c("barcode","SCanSNPID")),  by = "barcode") %>%
      dplyr::inner_join(dplyr::select(Demuxlet, c("barcode","Demuxlet_BEST")), by = "barcode")
  }
#head(mergedIDs)

#mergedIDs  <- merge(SoC, Demuxlet_V1,by = "barcode")
#mergedIDs  <- merge(mergedIDs, Vireo ,by = "barcode")
#mergedIDs  <- merge(mergedIDs, Scansnp ,by = "barcode")
#mergedIDs  <- merge(mergedIDs, Demuxlet ,by = "barcode")
```


****

## __3. Frequency of doublets__


### 3.1 __Demuxlet V1: Frequency of doublets__

```{r, collapse=TRUE}
table(mergedIDs$Demuxlet_V1_BEST)["doublet"]
round(prop.table(table(mergedIDs$Demuxlet_V1_BEST))*100, 2)

```

```{r, collapse=TRUE, fig.height=6, fig.width=6}
DemTypeDF <- data.frame("dblType" = factor(ifelse((mergedIDs$Demuxlet_V1_BEST != "doublet") & (mergedIDs$Demuxlet_V1_BEST != "unassigned"), "singlet", mergedIDs$Demuxlet_V1_BEST)))

DemType <- ggplot(data=DemTypeDF, aes(x=dblType, fill=dblType)) +
  geom_bar(alpha=0.65) +
  geom_text(stat='count', aes(label=..count..), vjust=1.2) +
  ggtitle('Demuxlet: Doublet Number') + ylab('Droplet Number') + xlab('Droplet Type') +
  scale_fill_hue(h=c(0, 120)) +
  theme_bw()

DemType
```

##### The percentage of __doublets__ detected __by Demuxlet V1 is `r round(table(mergedIDs$Demuxlet_V1_BEST)["doublet"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### The percentage of __unassigned__ detected __by Demuxlet V1 is `r round(table(mergedIDs$Demuxlet_V1_BEST)["unassigned"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### Total __non empty droplets__ in experiment is  __`r dim(mergedIDs)[1]`__



### 3.1b __Demuxlet V2: Frequency of doublets__

```{r, collapse=TRUE}
table(mergedIDs$Demuxlet_BEST)["doublet"]
round(prop.table(table(mergedIDs$Demuxlet_BEST))*100, 2)

```

```{r, collapse=TRUE, fig.height=6, fig.width=6}
DemTypeV2DF <- data.frame("dblType" = factor(ifelse((mergedIDs$Demuxlet_BEST != "doublet") & (mergedIDs$Demuxlet_BEST != "unassigned"), "singlet", mergedIDs$Demuxlet_BEST)))

DemTypeV2 <- ggplot(data=DemTypeV2DF, aes(x=dblType, fill=dblType)) +
  geom_bar(alpha=0.65) +
  geom_text(stat='count', aes(label=..count..), vjust=1.2) +
  ggtitle('Demuxlet V2: Doublet Number') + ylab('Droplet Number') + xlab('Droplet Type') +
  scale_fill_hue(h=c(0, 120)) +
  theme_bw()


DemTypeV2
```


##### The percentage of __doublets__ detected __by Demuxlet V2 is `r round(table(mergedIDs$Demuxlet_BEST)["doublet"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### The percentage of __unassigned__ detected __by Demuxlet V2 is `r round(table(mergedIDs$Demuxlet_BEST)["unassigned"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### Total __non empty droplets__ in experiment is  __`r dim(mergedIDs)[1]`__




### 3.2 __SoupOrCell: Frequency of doublets__

```{r, collapse=TRUE}
table(mergedIDs$SoCID)["doublet"]
round(prop.table(table(mergedIDs$SoCID))*100, 2)
```

```{r, collapse=TRUE, fig.height=4.5, fig.width=8}
SocDF <- data.frame("dblType" = factor(ifelse((mergedIDs$SoCID != "doublet") & (mergedIDs$SoCID != "unassigned"), "singlet", mergedIDs$SoCID)))

SocType <- ggplot(data=SocDF, aes(x=dblType, fill=dblType)) +
  geom_bar(alpha=0.65) +
  geom_text(stat='count', aes(label=..count..), vjust=1.2) +
  ggtitle('SoupOrCell: Doublet Number') + ylab('Droplet Number') + xlab('Droplet Type') +
  scale_fill_hue(h=c(0, 120)) +
  theme_bw()


SocType
```

##### The percentage of __doublets__ detected __by SoupOrCell is `r round(table(mergedIDs$SoCID)["doublet"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### The percentage of __unassigned__ detected __by SoupOrCell is `r round(table(mergedIDs$SoCID)["unassigned"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### Total __non empty droplets__ in experiment is  __`r dim(mergedIDs)[1]`__



### 3.3 __Vireo: Frequency of doublets__

```{r, collapse=TRUE}
table(mergedIDs$VireoID)["doublet"]
round(prop.table(table(mergedIDs$VireoID))*100, 2)

```

```{r, collapse=TRUE, fig.height=4.5, fig.width=8}
VireoDF <- data.frame("dblType" = factor(ifelse((mergedIDs$VireoID != "doublet") & (mergedIDs$VireoID != "unassigned"), "singlet", mergedIDs$VireoID)))

VireoType <- ggplot(data=VireoDF, aes(x=dblType, fill=dblType)) +
  geom_bar(alpha=0.65) +
  geom_text(stat='count', aes(label=..count..), vjust=1.2) +
  ggtitle('Vireo: Doublet Number') + ylab('Droplet Number') + xlab('Droplet Type') +
  scale_fill_hue(h=c(0, 120)) +
  theme_bw()


VireoType
```

##### The percentage of __doublets__ detected __by Vireo is `r round(table(mergedIDs$VireoID)["doublet"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### The percentage of __unassigned__ detected __by Vireo is `r round(table(mergedIDs$VireoID)["unassigned"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### Total __non empty droplets__ in experiment is  __`r dim(mergedIDs)[1]`__


### 3.4 __SCanSnp: Frequency of doublets__

```{r, collapse=TRUE}

table(mergedIDs$SCanSNPID)["doublet"]
round(prop.table(table(mergedIDs$SCanSNPID))*100, 2)

```

```{r, collapse=TRUE, fig.height=4.5, fig.width=8}
ScansnpDF <- data.frame("dblType" = factor(ifelse((mergedIDs$SCanSNPID != "doublet") & (mergedIDs$SCanSNPID != "unassigned"), "singlet", mergedIDs$SCanSNPID)))

ScansnpType <- ggplot(data=ScansnpDF, aes(x=dblType, fill=dblType)) +
  geom_bar(alpha=0.65) +
  geom_text(stat='count', aes(label=..count..), vjust=1.2) +
  ggtitle('SCanSNP: Doublet Number') + ylab('Droplet Number') + xlab('Droplet Type') +
  scale_fill_hue(h=c(0, 120)) +
  theme_bw()


ScansnpType
```


##### The percentage of __doublets__ detected __by SCanSNP is `r round(table(mergedIDs$SCanSNPID)["doublet"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### The percentage of __unassigned__ detected __by SCanSNP is `r round(table(mergedIDs$SCanSNPID)["unassigned"]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)`__
##### Total __non empty droplets__ in experiment is  __`r dim(mergedIDs)[1]`__


### 3.5 __scDblFinder: Frequency of doublets__

```{r, collapse=TRUE}

if (!is.null(FilteredFeaturesPath)){
table(mergedIDs$scDblFinder_type)["doublet"]
round(prop.table(table(mergedIDs$scDblFinder_type))*100, 2)
}

```

```{r, collapse=TRUE, fig.height=4.5, fig.width=8}
if (!is.null(FilteredFeaturesPath)){
scDblFinderDF <- data.frame("dblType" = factor(ifelse((mergedIDs$scDblFinder_type != "doublet") & (mergedIDs$scDblFinder_type != "unassigned"), "singlet", mergedIDs$scDblFinder_type)))

scDblFinder <- ggplot(data=scDblFinderDF, aes(x=dblType, fill=dblType)) +
  geom_bar(alpha=0.65) +
  geom_text(stat='count', aes(label=..count..), vjust=1.2) +
  ggtitle('scDblFinde: Doublet Number') + ylab('Droplet Number') + xlab('Droplet Type') +
  scale_fill_hue(h=c(0, 120)) +
  theme_bw()


scDblFinder
}


```

```{r, collapse=TRUE}
if (!is.null(FilteredFeaturesPath)){
  DBLpercentage <- round(table(mergedIDs$scDblFinder_type)["doublet"][[1]]/dim(mergedIDs)[1]*100,2) %>% replace_na(0)
  message <- paste0("The percentage of doublets detected by scDblFinder is ", as.character(DBLpercentage))
}else{
  message <- "Please provide Filtered Features path to Detect doublets via scDblFinder"
}
```


##### `r message`
##### Total __non empty droplets__ in experiment is  __`r dim(mergedIDs)[1]`__


****

## __4. Identity Assignment__

### 4.1 __Plot of Cells Assignments across tools__

```{r, echo = FALSE, warning  = FALSE}
## SOC ID assignment

Names <- as.vector(unique(mergedIDs$SoCID))
SoCNoDBLs <- Names[Names != "doublet" & Names != "unassigned" ]

DMX_IDs <- as.vector(unique(mergedIDs$Demuxlet_BEST)[(unique(mergedIDs$Demuxlet_BEST) != "doublet" & unique(mergedIDs$Demuxlet_BEST) != "unassigned")])


combinations <- permn(DMX_IDs)
BestCombOverlap <- 0
for(combination in seq(length(combinations))){
  mergedIDs_temp <- mergedIDs
  CombTemp <- combinations[[combination]]
  SoCNoDBLs_temp <- SoCNoDBLs



  for(pos in seq(length(CombTemp))){
        SoCNoDBLs_temp[[pos]] <- CombTemp[[pos]]
        mergedIDs_temp$SoCID[mergedIDs_temp$SoCID==SoCNoDBLs[[pos]]] <- CombTemp[[pos]]

    if (pos == length(CombTemp)){
      if (length(which(mergedIDs_temp$SoCID == mergedIDs_temp$Demuxlet_BEST)) > BestCombOverlap){
        mergedIDs_SoCMapped <- mergedIDs_temp
        BestCombOverlap <- length(which(mergedIDs_temp$SoCID == mergedIDs_temp$Demuxlet_BEST))
        SoCMappings <- list("SoCClusters"=SoCNoDBLs, "Mappings"=SoCNoDBLs_temp)

      }
    }
  }
}


BestCombOverlapRatio <- BestCombOverlap/length(mergedIDs$barcode)
mappingsDF <- as.data.frame(SoCMappings)

```
<font size="4"> SoC Identities are predicted by overlapping SoC clusters (singlets only) with Demuxlet V2 BEST.GUESS (singlets only)
Agreement was  `r round(BestCombOverlapRatio*100,2)`%</font>

```{r, echo = FALSE, warning  = FALSE}

mappingsDF

```

```{r, echo = FALSE, warning  = FALSE, fig.height = 10, fig.width = 16}
Software <- colnames(mergedIDs_SoCMapped)
Software <- Software[Software != "barcode"]

IDdf <- data_frame()
for (tool in Software){
  IDdf_temp<- as.data.frame(table(mergedIDs_SoCMapped[[tool]]), stringsAsFactors = F)
  colnames(IDdf_temp) <- c("ID","Cells")
  IDdf_temp$Software <- tool
  IDdf <- rbind(IDdf,IDdf_temp)
}

ggplot(data=IDdf, aes(x=Software, y=Cells, fill=ID)) +
  geom_bar(stat="identity", color="black",position=position_dodge())+
  theme_minimal()+
  theme(axis.text.x = element_text(size =  15,angle = 90), axis.text.y = element_text(size =  15), axis.title.y= element_text(size =  20), axis.title.x= element_text(size =  20))



```


### 4.2 __Pair-wise comparison__

#### 4.2.1 __Demuxlet vs SoC__


```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ Demuxlet_V1_BEST + SoCID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = "SoC clusters - Demuxlet overlaps", las=2)
```

```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ Demuxlet_BEST + SoCID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = "SoC clusters - Demuxlet overlaps", las=2)
```


#### 4.2.2 __Demuxlet vs Vireo__


```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ Demuxlet_BEST + VireoID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = "Vireo - Demuxlet  overlaps", las=2)
```

#### 4.2.3 __Demuxlet vs ScanSnp__


```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ Demuxlet_BEST + SCanSNPID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = " SCanSNP - Demuxlet  overlaps", las=2)
```


#### 4.2.4 __Vireo vs ScanSnp__


```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ VireoID + SCanSNPID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = "SCanSNP - Vireo overlaps", las=2)
```


#### 4.2.4 __Vireo vs SoC__


```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ VireoID + SoCID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = " SoC clusters - Vireo overlaps", las=2)
```


#### 4.2.5 __SCanSNP vs SoC__


```{r, echo = FALSE, warning  = FALSE, fig.height = 7.5, fig.width = 7.5}
mosaicplot(~ SCanSNPID + SoCID , data = mergedIDs_SoCMapped, col=c("darkolivegreen1", "darkgoldenrod1", "brown1", "darkorchid4", "cadetblue4","cyan2", "deepskyblue"), main = "SoC clusters - SCanSNP overlaps", las=2)
```


### 4.3 __Plot of Agreement between softwares__

```{r, echo = FALSE, warning  = FALSE, fig.height = 10, fig.width = 16}

Software_no_scDBLFinder <- Software[!Software %in% c("scDblFinder_type")]
SoftwareComb <- as.data.frame(combn(Software_no_scDBLFinder, 2), stringsAsFactors = F)

AgreementDF <- data_frame()
for (i in colnames(SoftwareComb)){
  Tool1 <- SoftwareComb[[i]][1]
  Tool2 <- SoftwareComb[[i]][2]
  CombEntry <- paste0(Tool1, "_VS_", Tool2)
  OverlapRatio <- round(length(which(mergedIDs_SoCMapped[[Tool1]] == mergedIDs_SoCMapped[[Tool2]]))/length(mergedIDs_SoCMapped[[Tool1]])*100,2)
  AgreementDF_temp <- as.data.frame(data_frame(Software = CombEntry, AgreementRatio = OverlapRatio))
  AgreementDF<- rbind(AgreementDF, AgreementDF_temp)
}



AgreementDF <- arrange(AgreementDF, AgreementRatio)
AgreementDF$Software <- factor(AgreementDF$Software, levels = AgreementDF$Software)



ggplot(data=AgreementDF, aes(x=Software, y = AgreementRatio,fill=Software)) +
  geom_bar(stat="identity", color="black",position=position_dodge())+
  theme_minimal()+
  coord_flip()+ theme(legend.position = "none")+
  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, 10))+
  theme(axis.text.x = element_text(size =  15), axis.text.y = element_text(size =  15), axis.title.y= element_text(size =  20), axis.title.x= element_text(size =  20))



```


## 5 __Counts Plot__

```{r, echo = FALSE, warning  = FALSE, fig.height = 40, fig.width = 16}

if (!is.null(FilteredFeaturesPath)){

  par(mfrow=c(5,1))

  sc.data <- Read10X(data.dir = FilteredFeaturesPath)
  if (is.list(sc.data)){
    SeuratObject <- CreateSeuratObject(counts = sc.data[["Gene Expression"]])
  } else {
    SeuratObject <- CreateSeuratObject(counts = sc.data)
  }
  rownames(mergedIDs_SoCMapped) <- mergedIDs_SoCMapped$barcode
  SeuratObject[["SoCID"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "SoCID"]
  SeuratObject[["Demuxlet_V1_BEST"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "Demuxlet_V1_BEST"]
  SeuratObject[["VireoID"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "VireoID"]
  SeuratObject[["SCanSNPID"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "SCanSNPID"]
  SeuratObject[["Demuxlet_BEST"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "Demuxlet_BEST"]
  SeuratObject[["scDblFinder_type"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "scDblFinder_type"]
  SeuratObject[["dropkick_label_03"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "dropkick_label_03"]

  SeuratObject$log_counts <- log10(SeuratObject$nCount_RNA+1)

plotList <- list()
plotList[["SoCIDplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "SoCID", pt.size = 0.5)+ labs(title = "Soup or Cell IDs")
plotList[["Demuxlet_V1_BESTplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "Demuxlet_V1_BEST", pt.size = 0.5)+ labs(title = "Demuxlet_V1 IDs")
plotList[["VireoIDplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "VireoID", pt.size = 0.5)+ labs(title = "Vireo IDs")
plotList[["SCanSNPIDplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "SCanSNPID", pt.size = 0.5)+ labs(title = "SCanSNP IDs")
plotList[["Demuxlet_BESTplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "Demuxlet_BEST", pt.size = 0.5)+ labs(title = "Demuxlet IDs")
plotList[["scDblFinder_typeplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "scDblFinder_type", pt.size = 0.5)+ labs(title = "scDblFinder_type IDs")
plotList[["dropkick"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "dropkick_label_03", pt.size = 0.5)+ labs(title = "dropkick type")

CombinePlots(plotList, ncol = 1, legend = NULL)

} else {
print("Please provide Filtered Features path for counts plot across IDs")
}

```





### 6 __Accuracy assessment__

```{r, echo = FALSE, warning  = FALSE, fig.height = 10, fig.width = 16}

if (!is.null(BarcodeMap)){

  BarcodeMap <- BarcodeMap

  accuracyPlotList <- list()
  SynthReferences <- read.csv(BarcodeMap, header = T, sep = "\t", stringsAsFactors = F , col.names = c("barcode","DBLstatus","ID"))


HOMODBLs<-as.data.frame(str_split_fixed(data.frame(barcode = SynthReferences$barcode, ID= SynthReferences$ID, stringsAsFactors = F)$ID, ",", 2), stringsAsFactors = F)
HOMODBLs$barcode <- data.frame(barcode = SynthReferences$barcode, ID= SynthReferences$ID, stringsAsFactors = F)$barcode
HOMODBLs<-HOMODBLs[HOMODBLs$V1 == HOMODBLs$V2,]$barcode


SynthReferences <- SynthReferences[! SynthReferences$barcode %in% HOMODBLs,]


SynthReferences$ID <- ifelse(SynthReferences$DBLstatus == "Singlet", SynthReferences$ID, "doublet")





mergedIDs_SoCMapped <- data.frame(lapply(mergedIDs_SoCMapped, function(x) {
                  gsub("Sample_S20568_CFGSH52D18", "CFG", x)
              }))

mergedIDs_SoCMapped <- data.frame(lapply(mergedIDs_SoCMapped, function(x) {
                  gsub("MIFF1DAY100REP1", "MIFF1", x)
              }))
mergedIDs_SoCMapped <- data.frame(lapply(mergedIDs_SoCMapped, function(x) {
                  gsub("809_1_5", "809", x)
              }))
mergedIDs_SoCMapped <- data.frame(lapply(mergedIDs_SoCMapped, function(x) {
                  gsub("KOLF2C1DAY100REP1", "KOLF", x)
              }))
mergedIDs_SoCMapped <- data.frame(lapply(mergedIDs_SoCMapped, function(x) {
                  gsub("3391BDAY100REP1", "3391B", x)
              }))

merge_Synth <- merge(mergedIDs_SoCMapped, SynthReferences, by = "barcode")




Software <- colnames(mergedIDs_SoCMapped)
Software <- Software[Software != "barcode"]


CorrectAssignsDF <- data_frame()
for (tool in Software){
  ToolID <- data_frame()
  for (IdentityTemp in unique(merge_Synth$ID)){
    ID_ss <- merge_Synth[merge_Synth$ID == IdentityTemp,]
    ID_Agreement<-length(which(ID_ss[[tool]] == ID_ss$ID))/length(ID_ss$ID)
    ToolID_temp <- as.data.frame(data_frame(Software = tool, Identity = IdentityTemp, CorrectAssigns = ID_Agreement))
    ToolID <- rbind(ToolID,ToolID_temp )
  }
  CorrectAssignsDF <- rbind(ToolID, CorrectAssignsDF)
}



accuracyPlotList[["AccuracyPerLine"]] <- ggplot(data=CorrectAssignsDF, aes(x="Software", y="CorrectAssigns", fill="Identity")) +
  geom_bar(stat="identity", position=position_dodge(0.6), width = 0.6)+
  scale_fill_brewer(palette="Paired")+
    labs(title="Accuracy Per ID")+
  theme_minimal()+
  theme(plot.title = element_text(hjust = 0.5))+
    theme(axis.text.x = element_text(size =  15), axis.text.y = element_text(size =  15), axis.title.y= element_text(size =  20), axis.title.x= element_text(size =  20), plot.title = element_text(size = 30, face = "bold"))




CorrectOverallAssignsDF <- data_frame()
for (tool in Software){
  OverallCorrectRatio <- length(which(merge_Synth[[tool]] == merge_Synth$ID))/length(merge_Synth$ID)*100
  CorrectOverallAssignsDF_temp  <- data.frame(CorrectOverallAssignsRatio = OverallCorrectRatio, Software = tool)
  CorrectOverallAssignsDF <- rbind(CorrectOverallAssignsDF, CorrectOverallAssignsDF_temp)

}


CorrectOverallAssignsDF <- arrange(CorrectOverallAssignsDF, CorrectOverallAssignsRatio)
CorrectOverallAssignsDF$Software <- factor(CorrectOverallAssignsDF$Software, levels = CorrectOverallAssignsDF$Software)




accuracyPlotList[["OverallAccuracy"]] <- ggplot(data=CorrectOverallAssignsDF, aes(x=Software, y = CorrectOverallAssignsRatio,fill=Software)) +
  geom_bar(stat="identity", color="black",position=position_dodge())+
  scale_fill_brewer(palette="Set3")+
  theme_minimal()+
  labs(title="Overall Accuracy")+
  theme(plot.title = element_text(hjust = 0.5))+
  coord_flip()+ theme(legend.position = "none")+
  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, 10))+
    theme(axis.text.x = element_text(size =  15), axis.text.y = element_text(size =  15), axis.title.y= element_text(size =  20), axis.title.x= element_text(size =  20), plot.title = element_text(size = 30, face = "bold"))

gridExtra::grid.arrange(grobs = accuracyPlotList, ncol =1,gp=gpar(fontsize=20))


} else {
print("Please provide BarcodeMap to assess the Deconvolution accuracy ")
}


```


### 7 __Best Agreement metrices calculation__

```{r , collapse=TRUE}


AgreementDF_noDv1 <- AgreementDF[grep("Demuxlet_V1_BEST", AgreementDF$Software, invert=TRUE),]
bestComb <- unlist(strsplit(as.character(AgreementDF_noDv1[order(AgreementDF_noDv1$AgreementRatio, decreasing = TRUE),"Software"][1]), "_VS_"))
mergedIDs_SoCMappedSS <- mergedIDs_SoCMapped[colnames(mergedIDs_SoCMapped)[!colnames(mergedIDs_SoCMapped) %in% c("barcode","Demuxlet_V1_BEST")]]
TotalIDsets <- list()
for (id in unique(as.vector(t(mergedIDs_SoCMappedSS)))) {
  TotalIDsets[[id]] <- c()
  for (tool in bestComb){
    tempVec <- ifelse(mergedIDs_SoCMapped[tool] == id, mergedIDs_SoCMapped$barcode, "remove")[ifelse(mergedIDs_SoCMapped[tool] == id, mergedIDs_SoCMapped$barcode, "remove") != "remove"]
    TotalIDsets[[id]] <- unique(c(TotalIDsets[[id]], tempVec))
  }
  TotalIDsets[[id]] <- length(TotalIDsets[[id]])
}
IDagreementRatios <- list()
for (id in unique(as.vector(t(mergedIDs_SoCMappedSS)))) {
  nTools <- length(bestComb)
  tempSS <- mergedIDs_SoCMappedSS[bestComb]
  tempSS[tempSS==id]<-1
  tempSS[tempSS!=1]<-0
  tempSS <- sapply( tempSS, as.numeric )
  IDagreementRatios[[paste0("shared_",id)]] <- round(length(which(rowSums(tempSS) == nTools))/TotalIDsets[[id]],2)
}
IDagreementRatios[["shared_overall"]] <- round(as.numeric(AgreementDF_noDv1[order(AgreementDF_noDv1$AgreementRatio, decreasing = TRUE),"AgreementRatio"][1])/100,2)
IDagreementRatios[["best_agreement"]] <- as.character(AgreementDF_noDv1[order(AgreementDF_noDv1$AgreementRatio, decreasing = TRUE),"Software"][1])
IDagreementRatiosDF <- as.data.frame(IDagreementRatios)


```


### 8 __scDblFinder vs genetic doublets detection__

```{r, echo = FALSE, warning  = FALSE, fig.height = 10, fig.width = 16}


if (!is.null(FilteredFeaturesPath)){

mergedIDs_SoCMapped_SNG <- mergedIDs_SoCMapped
rownames(mergedIDs_SoCMapped_SNG) <- mergedIDs_SoCMapped_SNG$barcode
mergedIDs_SoCMapped_SNG$barcode <- NULL

mergedIDs_SoCMapped_SNG[(mergedIDs_SoCMapped_SNG != "doublet") & (mergedIDs_SoCMapped_SNG != "unassigned")] <- "singlet"


Combs <- as.data.frame(combn(Software, 2), stringsAsFactors = F)

scDblCombs <- c()
for (col in colnames(Combs)){
  if ("scDblFinder_type" %in% c(Combs[col])[[1]]) {

    scDblCombs <-c(scDblCombs,col)
  }
}

Combs <- Combs[scDblCombs]




vennList <- list()
for (i in colnames(Combs)){

  Tool1 <- Combs[[i]][1]
  Tool2 <- Combs[[i]][2]

  combName <- paste0(Tool1,"_VS_",Tool2)

  DBLsBarcodeslists <- list()
  DBLsBarcodeslists[[Tool1]] <- rownames(mergedIDs_SoCMapped_SNG[mergedIDs_SoCMapped_SNG[Tool1] == "doublet",])
  DBLsBarcodeslists[[Tool2]] <- rownames(mergedIDs_SoCMapped_SNG[mergedIDs_SoCMapped_SNG[Tool2] == "doublet",])


  grid.newpage()
  venn_object <- venn.diagram(
          DBLsBarcodeslists,
          category.names = c(Tool1 , Tool2 ),
          # Circles
          lwd = 2,
          lty = 'blank',
          fill = c("#E69F00", "#009E73"),
          # Numbers
          cex = .9,
          fontface = "italic",
          main = combName,
          # Set names
          cat.cex = 1,
          cat.fontface = "bold",
          cat.default.pos = "outer",
          filename = NULL)
  grid.draw(venn_object)
}


} else {
  print("Please provide Filtered Features path to compare genetic DBLs vs Clustering DBLs")
}


```



## 9 __Exporting results__

```{r , collapse=TRUE}

if (is.null(BarcodeMap)){
  DStype <- "1"
} else {
  DStype <- "0"

}




DSsize <- dim(mergedIDs_SoCMapped)[1]

MetricsDF <- data.frame(matrix(, nrow=0, ncol=5))
colnames(MetricsDF) <- c("Software","DBLsRate","TotalDroplets","Dataset","DStype")

toolsList <- colnames(mergedIDs_SoCMapped)[!colnames(mergedIDs_SoCMapped) %in% c("barcode")]
IDsList <- unique(as.vector(t(mergedIDs_SoCMappedSS)))

for (tool in toolsList){
  ToolDBLsRatio <- as.numeric(round(table(mergedIDs_SoCMapped[tool])["doublet"]/DSsize*100,2))
  toolRow <- data.frame("Software"=tool, "DBLsRate"=ToolDBLsRatio, "TotalDroplets"= DSsize, "Dataset"=Dataset, "DStype"=DStype)
  toolRow <- cbind(toolRow,IDagreementRatiosDF )
  MetricsDF <- rbind(MetricsDF, toolRow)
}




if (DStype == "0" ){
  rownames(CorrectOverallAssignsDF) <- CorrectOverallAssignsDF$Software
  MetricsDF$SynthAccuracy <- CorrectOverallAssignsDF[MetricsDF$Software,"CorrectOverallAssignsRatio"]
  MockCountsDF <- data.frame(matrix(, nrow=length(toolsList), ncol=length(IDsList)))
  #AppendMock Counts matrix
  colnames(MockCountsDF) <- paste0("logCounts_",IDsList)
  MetricsDF <- cbind(MetricsDF, MockCountsDF)
} else {
  MetricsDF$SynthAccuracy <- NA
  countsPerToolDF <- data.frame(matrix(, nrow=0, ncol=length(IDsList)))
  colnames(countsPerToolDF) <- paste0("logCounts_",IDsList)
  for (tool in toolsList){
    countsPerTool <- SeuratObject@meta.data[c(tool,"log_counts")]
    countsPerTool <- aggregate(countsPerTool[,"log_counts"], list(countsPerTool[,tool]), median)
    rownames(countsPerTool) <- paste0("logCounts_",countsPerTool$Group.1)
    countsPerTool$Group.1 <- NULL
    countsPerTool <- as.data.frame(t(countsPerTool))
    rownames(countsPerTool) <- NULL
    countsPerToolDF <- bind_rows( countsPerToolDF,countsPerTool)
  }
  MetricsDF <- cbind(MetricsDF, countsPerToolDF)
}


write.table(MetricsDF, file=metricsFile, sep = "\t", row.names = F, col.names = T,quote = F)


```




## 10 __Multi call TSV preparation__

```{r , collapse=TRUE}


# SoC

if (!is.null(FilteredFeaturesPath)){
SoCSS <- read.csv(SoC_path, header = T, sep = "\t",stringsAsFactors=F)[,c("barcode","status","assignment")]
colnames(SoCSS) <- c("barcode","SoC_status","assignment")
SoCSS <- SoCSS %>% tidyr::separate(assignment, c('SoC_FirstID', 'SoC_SecondID'), sep="/")
SoCSS$SoC_FirstID <- plyr::mapvalues(SoCSS$SoC_FirstID, from = mappingsDF$SoCClusters , to = mappingsDF$Mappings)
SoCSS$SoC_SecondID <- plyr::mapvalues(SoCSS$SoC_SecondID, from = mappingsDF$SoCClusters , to = mappingsDF$Mappings)



# Vireo
VireoSS <- read.csv(Vireo_path, header = T, sep = "\t",stringsAsFactors=F) %>% dplyr::rename(barcode=cell, VireoID=donor_id)
VireoSS <-  VireoSS[,c("barcode","VireoID","best_singlet")]
colnames(VireoSS) <- c("barcode","Vireo_ID","Vireo_best_singlet")




# Demuxlet V2
DemuxletSS <- read.csv(Demuxlet_path, header = T, sep = "\t",stringsAsFactors=F)  %>% dplyr::rename(barcode=BARCODE) %>%
  tidyr::separate(BEST.GUESS, c('Demuxlet_FirstID', 'Demuxlet_SecondID','Prop'), sep=",")
#Demuxlet$Demuxlet_BEST <- ifelse(Demuxlet$DROPLET.TYPE != "SNG", Demuxlet$DROPLET.TYPE, Demuxlet$BEST.GUESS_1)
DemuxletSS[DemuxletSS=="DBL"]<-"doublet"
DemuxletSS[DemuxletSS=="AMB"]<-"unassigned"
DemuxletSS[DemuxletSS=="SNG"]<-"singlet"
DemuxletSS <- DemuxletSS[,c("barcode","DROPLET.TYPE", "Demuxlet_FirstID", "Demuxlet_SecondID")] %>% dplyr::rename(Demuxlet_dbltype =DROPLET.TYPE)



# ScansnpSS
ScansnpSS <- read.csv(SCanSNP_path, sep = "\t", header = T,stringsAsFactors=F) %>% dplyr::rename(ScansnpSSID=ID)
ScansnpSS$ScansnpSSID[ScansnpSS$ScansnpSSID == "notclassified"] <- "unassigned"
ScansnpSS[ScansnpSS=="Doublet"]<-"doublet"
ScansnpSS[ScansnpSS=="Singlet"]<-"singlet"
ScansnpSS$Quality <- ifelse((ScansnpSS$Quality != "LowQuality") & (ScansnpSS$Quality != "doublet"), "GoodQuality", ScansnpSS$Quality )
ScansnpSS <- ScansnpSS[,c("barcode","DropletType","FirstID","SecondID","Quality")]%>% dplyr::rename(ScansnpSS_Type = DropletType, ScansnpSS_FirstID=FirstID , ScansnpSS_SecondID = SecondID, ScansnpSS_Qual = Quality )



scDblFinderDF <- read.csv(scDblFinder_path, sep = "\t", header = F,stringsAsFactors=F, col.names = c("barcode","scDblFinder_type"))

dropkick <- read.csv(dropkick_path, sep = "\t", header = T,stringsAsFactors=F)

MultiCall <- SoCSS  %>%   dplyr::inner_join(VireoSS,  by = "barcode") %>%   dplyr::inner_join(ScansnpSS,  by = "barcode") %>% dplyr::inner_join(DemuxletSS, by = "barcode") %>%
    dplyr::inner_join(scDblFinderDF, by = "barcode") %>%    dplyr::inner_join(dropkick, by = "barcode")

  } else {

  print("Please provide Filtered Features To create aggregated ID call")
  }





```



### 10.1 __Multi call TSV Exploration__


```{r , collapse=TRUE}

if (!is.null(FilteredFeaturesPath)){

DBLsPerBC <- data.frame(barcode=MultiCall$barcode)
DBLsPerBC["DBLs_Call"] <- rowSums(MultiCall[c("SoC_status","Vireo_ID","ScansnpSS_Type","Demuxlet_dbltype")] == "doublet")



#Initilize empty DFs
Consensus <- data.frame(barcode=MultiCall$barcode)
Consensus_min1 <- data.frame(barcode=MultiCall$barcode)
Consensus_max2 <- data.frame(barcode=MultiCall$barcode)



for (id in unique(unlist(MultiCall[c("SoC_FirstID","Vireo_best_singlet","ScansnpSS_FirstID","Demuxlet_FirstID")]))){

  #Adding tool-specific score (1.5 for demuxlet/scansnp and 1 for Soc/Vireo) since in this phase we want to exploit demuxlet/scansnp sensitivity
  CallsPerBC <- as.data.frame(MultiCall[c("SoC_FirstID","Vireo_best_singlet","ScansnpSS_FirstID","Demuxlet_FirstID")] == id)
  CallsPerBC["SoC_FirstID"] <- ifelse(CallsPerBC[,"SoC_FirstID"], 1, 0)
  CallsPerBC["Vireo_best_singlet"] <- ifelse(CallsPerBC[,"Vireo_best_singlet"], 1, 0)
  CallsPerBC["ScansnpSS_FirstID"] <- ifelse(CallsPerBC[,"ScansnpSS_FirstID"], 1.5 , 0)
  CallsPerBC["Demuxlet_FirstID"] <- ifelse(CallsPerBC[,"Demuxlet_FirstID"], 1.5, 0)

  CallsPerBC[paste0("Calls_",id)] <- rowSums(CallsPerBC)

  Consensus[paste0("Consensus_",id)] <- as.vector(ifelse(CallsPerBC[paste0("Calls_",id)] >= 3, 1,0))


  #CallsPerBC[paste0("Calls_",id)] <- rowSums(MultiCall[c("SoC_FirstID","Vireo_best_singlet","ScansnpSS_FirstID","Demuxlet_FirstID")] == id)
  #Consensus[paste0("Consensus_",id)] <- as.vector(ifelse(CallsPerBC[paste0("Calls_",id)] >= 3, 1,0))
  #Consensus_min1[paste0("Consensus_",id)] <- as.vector(ifelse(CallsPerBC[paste0("Calls_",id)] >= 3, 1,0))
}




#LowQuality Assignment

LowQuality <- MultiCall[((MultiCall$ScansnpSS_Qual == "LowQuality") & (rowSums(Consensus[-1]) == 0)) | ((rowSums(Consensus[-1]) == 0) & (MultiCall$dropkick_label_03 == "LowQuality")),"barcode"]




#Genetic Doublets assignment
Genetic_DBLs <-  MultiCall[

(

((MultiCall$ScansnpSS_Type == "doublet") & ((MultiCall$SoC_status == "doublet") | (MultiCall$Vireo_ID == "doublet") | (MultiCall$Demuxlet_dbltype == "doublet") | (MultiCall$scDblFinder_type == "doublet") ))   |
((MultiCall$SoC_status == "doublet") & (MultiCall$Vireo_ID == "doublet") & (MultiCall$Demuxlet_dbltype == "doublet") &  (rowSums(Consensus[-1]) == 0))

)

&     (! MultiCall$barcode %in% LowQuality)



                           ,"barcode"]



#Transcriptomic Doublets assignment
Transcriptomic_DBLs <- MultiCall[(MultiCall$scDblFinder_type == "doublet") & (! MultiCall$barcode %in% Genetic_DBLs) & (! MultiCall$barcode %in% LowQuality),"barcode"]



SafeSNGs <- MultiCall[as.vector((rowSums(Consensus[-1]) == 1) & (!MultiCall$barcode %in% LowQuality) & (!MultiCall$barcode %in% Transcriptomic_DBLs) & (!MultiCall$barcode %in% Genetic_DBLs))
,"barcode"]


#Finally cells not meeting consensus and not falling into previous categories
LowQuality <- c(LowQuality, MultiCall[(!MultiCall$barcode %in% SafeSNGs) & (!MultiCall$barcode %in% LowQuality) & (!MultiCall$barcode %in% Transcriptomic_DBLs) & (!MultiCall$barcode %in% Genetic_DBLs),"barcode"])






#DF slices prep
SafeSNGsDF <- data.frame(barcode = SafeSNGs , Consensus = sapply(strsplit(colnames(Consensus[-1])[max.col(Consensus[Consensus$barcode %in% SafeSNGs,-1])],"_"), `[`, 2))
doubletDF <- data.frame(barcode = unique(c(Genetic_DBLs,Transcriptomic_DBLs))  , Consensus = "doublet")
LowQualDF <- data.frame(barcode = LowQuality , Consensus = "LowQuality")
#RecoveredSNGsDF <- data.frame(barcode = RecoveredSNGs , Consensus = sapply(strsplit(colnames(CallsPerBC[-1])[max.col(CallsPerBC[CallsPerBC$barcode %in% RecoveredSNGs,-1])],"_"), `[`, 2))

#FinalDF
ConsensusDF <- rbind(rbind(SafeSNGsDF,doubletDF), LowQualDF)

write.table(ConsensusDF,aggregatedCall, row.names = F, quote = F, sep = "\t")

rownames(ConsensusDF) <- ConsensusDF$barcode
  } else {

  print("Please provide Filtered Features To create aggregated ID call")
  }




```




### 11 __Counts Plot with aggregated Call__



```{r, echo = FALSE, warning  = FALSE, fig.height = 40, fig.width = 16}


if (!is.null(FilteredFeaturesPath)){

  par(mfrow=c(5,1))

  sc.data <- Read10X(data.dir = FilteredFeaturesPath)
  if (is.list(sc.data)){
    SeuratObject <- CreateSeuratObject(counts = sc.data[["Gene Expression"]])
  } else {
    SeuratObject <- CreateSeuratObject(counts = sc.data)
  }
  rownames(mergedIDs_SoCMapped) <- mergedIDs_SoCMapped$barcode
  SeuratObject[["SoCID"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "SoCID"]
  SeuratObject[["Demuxlet_V1_BEST"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "Demuxlet_V1_BEST"]
  SeuratObject[["VireoID"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "VireoID"]
  SeuratObject[["SCanSNPID"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "SCanSNPID"]
  SeuratObject[["Demuxlet_BEST"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "Demuxlet_BEST"]
  SeuratObject[["scDblFinder_type"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "scDblFinder_type"]
  SeuratObject[["dropkick_label_03"]] <- mergedIDs_SoCMapped[rownames(SeuratObject@meta.data), "dropkick_label_03"]
  SeuratObject[["AggregatedCall"]] <- ConsensusDF[rownames(SeuratObject@meta.data), "Consensus"]

  SeuratObject$log_counts <- log10(SeuratObject$nCount_RNA+1)

plotList <- list()
plotList[["SoCIDplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "SoCID", pt.size = 0.5)+ labs(title = "Soup or Cell IDs")
plotList[["Demuxlet_V1_BESTplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "Demuxlet_V1_BEST", pt.size = 0.5)+ labs(title = "Demuxlet_V1 IDs")
plotList[["VireoIDplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "VireoID", pt.size = 0.5)+ labs(title = "Vireo IDs")
plotList[["SCanSNPIDplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "SCanSNPID", pt.size = 0.5)+ labs(title = "SCanSNP IDs")
plotList[["Demuxlet_BESTplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "Demuxlet_BEST", pt.size = 0.5)+ labs(title = "Demuxlet IDs")
plotList[["scDblFinder_typeplot"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "scDblFinder_type", pt.size = 0.5)+ labs(title = "scDblFinder_type IDs")
plotList[["dropkick"]] <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "dropkick_label_03", pt.size = 0.5)+ labs(title = "dropkick IDs")
plotList[["AggregatedCall_plot"]]  <-  VlnPlot(SeuratObject, features = "log_counts", group.by = "AggregatedCall", pt.size = 0.5)+ labs(title = "AggregatedCall IDs")

CombinePlots(plotList, ncol = 1, legend = NULL)

} else {
print("Please provide Filtered Features path for counts plot across IDs")
}



```



### 12 __Plot of Cells Assignments across tools with aggregated Call__


```{r, echo = FALSE, warning  = FALSE, fig.height = 10, fig.width = 16}

if (!is.null(FilteredFeaturesPath)){

IDdf_temp<- as.data.frame(table(ConsensusDF$Consensus), stringsAsFactors = F)
colnames(IDdf_temp) <- c("ID","Cells")
IDdf_temp$Software <- "aggregatedCall"
IDdf <- rbind(IDdf,IDdf_temp)


ggplot(data=IDdf, aes(x=Software, y=Cells, fill=ID)) +
  geom_bar(stat="identity", color="black",position=position_dodge())+
  theme_minimal()+
  theme(axis.text.x = element_text(size =  15,angle = 90), axis.text.y = element_text(size =  15), axis.title.y= element_text(size =  20), axis.title.x= element_text(size =  20))

} else {
print("Please provide Filtered Features path for counts plot across IDs")
}

```
